### PR TITLE
Binary model support

### PR DESCRIPTION
**Pull request by Janaka-Steph**
Support Docker for web and binaries for desktop.
Introduce a new Service Controller class in charge of managing Prem Services.

The manifest for binaries contains the huggingFaceId and the files to download.
For now we can download (resumable) Tabby from HuggingFace, GGML files are chosen. We don't use Tabby cli anymore, the goal is to have a generic solution for any model.
One minor issue is adding isDownloading state to daemon, so that we prevent downloading multiple times. Code could be front-end side but cleaner if server side.
Next step is to serve/inference them with a Rust server.

Binaries:
Services are pulled from Rust and set in Tauri store.
When downloading and starting, you can check the log file at $APP_DIR/binaries/$SERVICE_NAME/$SERVICE_NAME.log

### CODE REVIEW

**Comment by filopedraz** [comment]
Nice, I'll have a deeper look at it soon. I suggest that for testing you should already create a PR to the [registry](https://github.com/premAI-io/prem-registry) with [Tabby](https://github.com/TabbyML/tabby/releases/tag/v0.1.1) service.

**Comment by Janaka-Steph** [comment]
> Nice, I'll have a deeper look at it soon. I suggest that for testing you should already create a PR to the [registry](https://github.com/premAI-io/prem-registry) with [Tabby](https://github.com/TabbyML/tabby/releases/tag/v0.1.1) service.

Is it "Tabby SantaCoder 1B" ?

File: src-tauri/src/controller_docker.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    daemon: App,
+}
+
+#[derive(Deserialize, Debug)]
+struct Config {
+    prem: Prem,
+}
+
+#[tauri::command]
+pub fn start_container() {
```
Comment (tiero) [inline src-tauri/src/controller_docker.rs:24]: we don't really want to support Docker going forward, but I see how this is helpful to have a "smooth" migration.

Comment (Janaka-Steph) [reply src-tauri/src/controller_docker.rs:24]: Yes, Docker still works like before. It doesn't cost much to keep it.

Comment (filopedraz) [reply src-tauri/src/controller_docker.rs:24]: Yep, and the petals services will need Docker for a while :)

Comment (tiero) [reply src-tauri/src/controller_docker.rs:24]: Since we decided we want to remove the Docker Wall ie. onboarding page, I think to re-enable docker models in a desktop context, we should figure out a different UI **inside** the service details page that willl do the docker checks, runs the prem daemon etc..

I think it ok to keep it for now, but will be not used this rust code unless we also change the UX

Comment (Janaka-Steph) [reply src-tauri/src/controller_docker.rs:24]: It's true but we could also keep the UI as is for a while. The only thing is that if user only wants to run a binary he will run Docker for nothing. But that strategy could ensure smooth migration.

Comment (filopedraz) [reply src-tauri/src/controller_docker.rs:24]: Yep, makes sense. Maybe from `Settings`? We can have `Enable Docker` with some tooltips explaining that is necessary Docker Desktop as a dependency. 

The button will check if Docker is installed and then run `premd`? Or we can have Tauri directly running the docker images?

**Comment by filopedraz** [comment]
Looks very nice @Janaka-Steph. Great job

File: src-tauri/src/controller_docker.rs (hunk 1 above)
Comment (tiero) [reply src-tauri/src/controller_docker.rs:24]: As we discussed offline with @filopedraz supporting docker it will require more work on support multiple use cases, hence complexity.
Since for providing blocks for petals we install already in the global user space, we can remove Docker entirely.

Comment (filopedraz) [reply src-tauri/src/controller_docker.rs:24]: Yep, I confirm

**Comment by filopedraz** [comment]
@swarnimarun please have a look as well.

File: src-tauri/src/main.rs [hunk 2]
```diff
@@ -1,123 +1,86 @@
...
+    name: Option<String>,
+    #[serde(rename = "needsUpdate")]
+    needs_update: Option<bool>,
+    #[serde(rename = "promptTemplate")]
+    prompt_template: Option<String>,
+    running: Option<bool>,
+    #[serde(rename = "runningPort")]
+    running_port: Option<u32>,
+    supported: Option<bool>,
+    r#type: Option<String>, //"binary" | "process",
```
Comment (swarnimarun) [inline src-tauri/src/main.rs:57]:
```rs
#[serde(rename = "type")]
ty: Option<String>, //"binary" | "process",
```

Raw-identifiers are generally only suitable for supporting backwards compatibility, or ffi use cases. We should avoid using them in production code otherwise.

Also raw-identifiers are a quite flaky part of Rust, they basically only exist at tokenization and are replaced before actual parsing which means if you try to using raw identifiers in unexpected position or non supported idents it may fail.

[Example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b58511329b10970290096c7de7e61563)

**Review by swarnimarun — COMMENTED** [review]
Also unwraps seem over used which I think we should avoid. Can submit a quick patch for it let me know if that's acceptable, or should I instead submit them as code suggestions?

File: src-tauri/src/main.rs (hunk 2 above)
Comment (Janaka-Steph) [reply src-tauri/src/main.rs:57]: Please submit code suggestions because I am still working on this PR

File: src-tauri/src/controller_binaries.rs [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+                is_downloading: false,
+                is_completed: true,
+                progress: 100.0,
+            },
+        );
+        return Ok(());
+    }
+
+    // Check if we're already downloading this file
+    println!("Checking if {:?} is already downloading", binary_path);
```
Comment (swarnimarun) [inline src-tauri/src/controller_binaries.rs:53]:
```suggestion
    println!("Checking if {} is already downloading", binary_path.display());
```

File: src-tauri/src/controller_binaries.rs [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        if let Some(status) = status_lock.get(&service_id) {
+            println!("Status: {:?}", status);
+            if status.is_downloading {
+                return Ok(());
+            }
+        }
+    }
+
+    // Update status to "downloading"
+    println!("Updating {:?} status to downloading", binary_path);
```
Comment (swarnimarun) [inline src-tauri/src/controller_binaries.rs:65]:
```suggestion
    println!("Updating {} status to downloading", binary_path.display());
```

File: src-tauri/src/controller_binaries.rs [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+        {
+            let mut status_lock = state.download_statuses.lock().await;
+            if let Some(status) = status_lock.get_mut(&service_id) {
+                status.progress = progress;
+            }
+        }
+    }
+
+    // Set execute permission
+    println!("Setting execute permission on {:?}", binary_path);
```
Comment (swarnimarun) [inline src-tauri/src/controller_binaries.rs:108]:
```suggestion
    println!("Setting execute permission on {}", binary_path.display());
```

File: src-tauri/src/controller_binaries.rs [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+            log_file
+                .try_clone()
+                .map_err(|e| format!("Failed to clone log file: {}", e))?,
+        ))
+        .stderr(std::process::Stdio::from(log_file))
+        .spawn()
+        .map_err(|e| format!("Failed to spawn child process: {}", e))?;
+
+    // Mark as done downloading
+    println!("Updating {:?} status to not downloading", binary_path);
```
Comment (swarnimarun) [inline src-tauri/src/controller_binaries.rs:142]:
```suggestion
    println!("Updating {} status to not downloading", binary_path.display());
```

File: src-tauri/src/controller_binaries.rs [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        .join(&service_id);
+    let binary_path = PathBuf::from(&service_dir).join(&service_id);
+    let log_path = PathBuf::from(&service_dir).join(format!("{}.log", service_id));
+
+    // Create service directory if it doesn't exist
+    std::fs::create_dir_all(&service_dir).map_err(|e| e.to_string())?;
+
+    // Check if file already exists
+    // TODO: If file exists we consider it is fully downloaded
+    println!("Checking if {:?} exists", binary_path);
```
Comment (swarnimarun) [inline src-tauri/src/controller_binaries.rs:30]:
```suggestion
    println!("Checking if {} exists", binary_path.display());
```

**Review by swarnimarun — COMMENTED** [review]
tldr; display renders non-utf8 path strings correctly.

long story.

A [Path](https://doc.rust-lang.org/stable/std/path/struct.Path.html) might contain non-Unicode data. This may perform lossy conversion, depending on the platform.

Where as Debug will correctly escape things it things are not utf-8 and will produce an unreadable mess in case of non utf-8 paths. (should work fine on most unix platforms, but windows and some others do like utf-16)

**Comment by filopedraz** [comment]
For simplicity, let's remove the Docker controller @Janaka-Steph. I thought about it extensively and we can get rid of that already now in Prem Desktop App.

**Comment by filopedraz** [comment]
On the other hand, @casperdcl if we don't manage to have petals services as binaries, we need to run them as a bash script similar to what we do with `Swarm Mode`.

File: src-tauri/src/controller_binaries.rs [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+        );
+    }
+
+    Downloader::new(hugging_face_id, service_id, service_dir, window)
+        .download_ggml_files()
+        .await
+}
+
+/*
+#[tauri::command(async)]
```
Comment (filopedraz) [inline src-tauri/src/controller_binaries.rs:56]: Remove commented code

File: src-tauri/src/download.rs [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+            Retry::spawn(
+                ExponentialBackoff::from_millis(100).map(jitter).take(2),
+                move || {
+                    self.download_file(
+                        self.registry
+                            .build_download_url(&self.hugging_face_id, file.as_ref()),
+                        format!("{}/{}", self.service_dir, file.as_ref()),
+                    )
+                },
+            )
```
Comment (swarnimarun) [inline src-tauri/src/download.rs:107]:
```suggestion
//             Retry::spawn(
//                 ExponentialBackoff::from_millis(100).map(jitter).take(2),
//                move || {
                    self.download_file(
                        self.registry
                            .build_download_url(&self.hugging_face_id, file.as_ref()),
                        format!("{}/{}", self.service_dir, file.as_ref()),
                    )
//                },
//            )
```
